### DISCUSSION TITLE
OracleStoredProcedureOperator & ORA- Errors & XCom

### DISCUSSION THREAD

**Discussion by kolfild26** [apache/airflow#23787] [Q&A] [answered]
Hi, 
I'm using `OracleStoredProcedureOperator` and I'd like to know how to retrieve the oracle stored procedure exit code in purpose of using it in dag.

As far as I see the airflow task only returns 0 when the prodedured's successfully completed and 1 when it's not no matter what causes its fail.  Well, we have the output text in log, like:
```
cx_Oracle.DatabaseError: ORA-06502: PL/SQL: numeric or value error: character to number conversion error
ORA-06512: at line 1
{local_task_job.py:154} INFO - Task exited with return code 1
```
but, we cannot use it in the following dag logic, because that error is not passed to XComs's return_value.

Some other airflow operators do this. For example, `SSHOperator` writes the output to XCom, if we `echo` smth in `command` parameter. And we can `echo $?` a thus we get the exit code of ssh command in the XCom variable.

- Maybe I miss smth and could anyone guide me how to achive what I need?
- If I'm right, is it possible to implement such functionality in `OracleStoredProcedureOperator`?
- Or is there any workaround of how to send `ORA-` exit code to XCom?

**Comment by kolfild26** [comment]
the question regards exit codes not the `OUT` parameters.
 the `OUT` parameters work, but they are only being passed to XCom in case of the successful finish.

**Reply by potiuk** [reply]
I think you need to look at the source code and figure out if you can improve it simply :D. It's open-source so anyone can provide a PR adding it. It's really operator-by-operator decision, so I guess anyone trying to answer that question would just have to do it - look up the code.

**Reply by potiuk** [reply]
This might be way faster than waiting for someone who might know the answer for that particular operator (there might be none who knows it by heart).

**Comment by kolfild26** [comment, marked as answer]
Thanks @potiuk. Done.
For those who might be also interested in such improvement.
DBApiHook has no any handler in case when cur.execute fails.
```
    def _run_command(self, cur, sql_statement, parameters):
        """Runs a statement using an already open cursor."""
        self.log.info("Running statement: %s, parameters: %s", sql_statement, parameters)
        if parameters:
            cur.execute(sql_statement, parameters)
        else:
            cur.execute(sql_statement)

        # According to PEP 249, this is -1 when query result is not applicable.
        if cur.rowcount >= 0:
            self.log.info("Rows affected: %s", cur.rowcount)
```
But as DbApiHook is a general class for all DB types it not recommended to change it there. Fix or inherit `OracleStoredProcedureOperator`

Instead of:
```
    def execute(self, context: 'Context') -> Optional[Union[List, Dict]]:
        self.log.info('Executing: %s', self.procedure)
        hook = OracleHook(oracle_conn_id=self.oracle_conn_id)
        return hook.callproc(self.procedure, autocommit=True, parameters=self.parameters)
```
do this:
```
    def execute(self, context: 'Context') -> Optional[Union[List, Dict]]:
        self.log.info('Executing: %s', self.procedure)
        hook = OracleHook(oracle_conn_id=self.oracle_conn_id)
        ti=context.get('task_instance')

        try:
            return hook.callproc(self.procedure, autocommit=True, parameters=self.parameters)
        except cx_Oracle.DatabaseError as e:
            code, mesg = e.args[0].message[:-1].split(': ', 1)
            ti.xcom_push(key='ORA', value=int(code.split('-')[1]))
            raise AirflowException(f"Task failed with exit code: {code}")
```
Thus in a fail scenario ORA return code wil appear in XCOM like this:
![изображение](https://user-images.githubusercontent.com/12990695/169669617-32ff6063-c0ac-4cf9-806c-1ed9f211989a.png)

**Reply by potiuk** [reply]
Maybe you would like to contribute it back @kolfild26 ? This is an easy way to become one of more than 2000 contirbutors. This is actually how Airflow is built.

**Reply by kolfild26** [reply]
Ok, will try a bit later

**Reply by kolfild26** [reply]
https://github.com/apache/airflow/pull/27319
